app/endpoints/nwb_validation.py

```diff
@@ -0,0 +1,365 @@
+
+
+def _handle_empty_file(file: UploadFile) -> NoReturn:
+    """Handle empty file upload by raising an appropriate HTTPException."""
+    L.error(f"Empty file uploaded: {file.filename}")
+    raise HTTPException(
+        status_code=HTTPStatus.BAD_REQUEST,
+        detail={
+            "code": ApiErrorCode.BAD_REQUEST,
+            "detail": "Uploaded file is empty",
+        },
+    )
+
+
+# New helper function to abstract the raise statement (TRY301 fix)
+def _handle_file_too_large(temp_path: str) -> NoReturn:
+    """Handles cleanup and raises FileTooLargeError when the file size limit is exceeded."""
+    pathlib.Path(temp_path).unlink(missing_ok=True)
+    max_mb = MAX_FILE_SIZE / (1024 * 1024)
+    msg = f"File size exceeds the limit of {max_mb:.0f} MB"
+    raise FileTooLargeError(msg)
+
+
+def _save_upload_to_tempfile(file: UploadFile, suffix: str) -> str:
+    """Save UploadFile to a temporary file synchronously."""
+    chunk_size = 1024 * 1024  # 1 MB
+    total_size = 0  # Track total size written
+
+    with tempfile.NamedTemporaryFile(delete=False, suffix=suffix) as temp_file:
+        temp_path = temp_file.name
+
+        try:
+            file.file.seek(0)  # Reset pointer
+            while True:
+                # Use chunk_size
+                chunk = file.file.read(chunk_size)
+                if not chunk:
+                    break
+
+                chunk_len = len(chunk)
+                total_size += chunk_len
+
+                # Check size limit before writing
+                if total_size > MAX_FILE_SIZE:
+                    _handle_file_too_large(temp_path)
+
+                temp_file.write(chunk)
+        except Exception:
+            if pathlib.Path(temp_path).exists():
+                pathlib.Path(temp_path).unlink(missing_ok=True)
+            raise
+        else:
+            return temp_path
+
+
+def _cleanup_temp_file(temp_path: str) -> None:
+    """Background task to clean up temporary file."""
+    if temp_path and pathlib.Path(temp_path).exists():
+        try:
+            pathlib.Path(temp_path).unlink()
+            L.debug(f"Cleaned up temp file: {temp_path}")
+        except OSError as e:
+            L.warning(f"Failed to delete temp NWB file: {e}")
+
+
+# Endpoint logic is now a standalone function
+def validate_nwb_file(
+    file: Annotated[UploadFile, File(description="NWB file to upload (.nwb)")],
+    background_tasks: BackgroundTasks,
+) -> NWBValidationResponse:
+    """Validates an uploaded .nwb file using registered readers."""
+    file_extension = pathlib.Path(file.filename).suffix.lower() if file.filename else ""
+    if file_extension != ".nwb":
+        raise HTTPException(
+            status_code=HTTPStatus.BAD_REQUEST,
+            detail={
+                "code": ApiErrorCode.BAD_REQUEST,
+                "detail": "Invalid file extension. Must be .nwb",
+            },
+        )
```

Comment:
You can also check `file.size` just after checking the extension, and fail fast if it's greater than the allowed size.

Reply:
This change has been implemented.